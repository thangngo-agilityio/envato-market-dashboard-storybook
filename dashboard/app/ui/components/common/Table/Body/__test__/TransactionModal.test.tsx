import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { TransactionModal } from '..';
import userEvent from '@testing-library/user-event';

// jest.mock('react-hook-form', () => ({
//   ...jest.requireActual('react-hook-form'),
//   useForm: () => ({
//     clearErrors: mockClearErrors,
//     formState: { isDirty: false },
//     handleSubmit: jest.fn(),
//     reset: jest.fn(),
//     control: {},
//   }),
// }));

describe('Transaction Modal', () => {
  const mockTransaction = {
    id: '123',
    customer: {
      firstName: 'John',
      lastName: 'Doe',
      email: 'doe@gmail.com',
      avatar: '',
      address: {
        street: '123 Main St',
        state: 'CA',
        city: 'City',
        zip: 12345,
      },
    },
  };

  it('match to snapshot', () => {
    const { container } = render(
      <TransactionModal isDelete transaction={mockTransaction} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render delete confirmation correctly', async () => {
    render(<TransactionModal isDelete transaction={mockTransaction} />);

    // Ensure delete confirmation message is rendered
    expect(
      screen.getByText(/Are you sure delete the transaction with id/i),
    ).toBeInTheDocument();
  });

  it('should call onDeleteTransaction on delete button click', async () => {
    const mockOnDeleteTransaction = jest.fn();
    render(
      <TransactionModal
        isDelete
        transaction={mockTransaction}
        onDeleteTransaction={mockOnDeleteTransaction}
      />,
    );

    // Click the delete button
    fireEvent.click(screen.getByText('Delete'));

    // Ensure onDeleteTransaction is called
    expect(mockOnDeleteTransaction).toHaveBeenCalled();
  });

  it('should call onCloseModal on cancel button click', async () => {
    const mockOnCloseModal = jest.fn();
    render(
      <TransactionModal
        isDelete
        transaction={mockTransaction}
        onCloseModal={mockOnCloseModal}
      />,
    );

    // Click the cancel button
    fireEvent.click(screen.getByText('Cancel'));

    waitFor(() => {
      // Ensure onCloseModal is called
      expect(mockOnCloseModal).toHaveBeenCalled();
    });
  });

  it('handleSubmitForm should update transaction, reset form, and close modal', async () => {
    const onUpdateTransactionMock = jest.fn();
    const onCloseModalMock = jest.fn();

    render(
      <TransactionModal
        transaction={mockTransaction}
        isDelete={false}
        onUpdateTransaction={onUpdateTransactionMock}
        onCloseModal={onCloseModalMock}
      />,
    );

    const inputField = screen.getByTestId('edit-field-name');

    await fireEvent.change(inputField, { target: { value: 'NewValue' } });

    await userEvent.click(screen.getByTestId('submit-transaction-form'));

    // Assertions
    waitFor(() => {
      expect(onUpdateTransactionMock).toHaveBeenCalled();
      expect(onCloseModalMock).toHaveBeenCalled();
    });
  });

  // it('should call handleChangeValue correctly', () => {
  //   const mockChangeHandler = jest.fn();

  //   render(<TransactionModal transaction={mockTransaction} />);

  //   // Find the input field
  //   const inputField = screen.getByTestId('edit-field-name');

  //   // Simulate a user typing in the input field
  //   fireEvent.change(inputField, { target: { value: 'NewValue' } });

  //   waitFor(() => {
  //     expect(mockClearErrors).toHaveBeenCalled();

  //     expect(mockChangeHandler).toHaveBeenCalled();
  //   });
  // });
});
