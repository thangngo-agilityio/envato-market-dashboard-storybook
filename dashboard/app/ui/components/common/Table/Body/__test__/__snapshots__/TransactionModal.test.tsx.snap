// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Modal match to snapshot 1`] = `
<div>
  <div
    class="css-0"
  >
    <p
      class="chakra-text css-1ezsviu"
    >
      Are you sure delete the transaction with id:
      <span
        class="chakra-text css-16ximt"
      >
        123
      </span>
      ?
    </p>
    <div
      class="css-ofbpde"
    >
      <button
        class="chakra-button css-nz6i3s"
        data-testid="accept-del"
        type="button"
      >
        Delete
      </button>
      <button
        class="chakra-button css-vxb15r"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
